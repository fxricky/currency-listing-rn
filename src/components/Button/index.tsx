import colors from "@/themes/colors";
import React, { useEffect, useState } from "react";
import {
  Pressable,
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  ViewStyle,
} from "react-native";

type Props = {
  label: string;
  variant?: "normal" | "danger" | "success" | "focus";
  onPress: () => void;
};

type VariantStyles = {
  container: StyleProp<ViewStyle>;
  txtLabel: StyleProp<TextStyle>;
};

export default function Button({
  label,
  variant = "normal",
  onPress,
}: Props): React.ReactElement {
  const [variantStyles, setVariantStyles] = useState<VariantStyles>(
    getStyles(variant),
  );

  useEffect(() => {
    setVariantStyles(getStyles(variant));
  }, [variant]);

  return (
    <Pressable
      onPress={onPress}
      style={[styles.baseContainer, variantStyles.container]}>
      <Text
        style={[styles.baseLabel, variantStyles.txtLabel]}
        numberOfLines={1}>
        {label}
      </Text>
    </Pressable>
  );
}

function getStyles(variant: Props["variant"]): VariantStyles {
  switch (variant) {
    case "danger":
      return {
        container: styles.btnDanger,
        txtLabel: styles.txtDanger,
      };
    case "success":
      return {
        container: styles.btnSuccess,
        txtLabel: styles.txtSuccess,
      };

    case "focus":
      return {
        container: styles.btnFocus,
        txtLabel: styles.txtFocus,
      };
    default:
      return {
        container: styles.btnNormal,
        txtLabel: styles.txtNormal,
      };
  }
}

const styles = StyleSheet.create({
  baseContainer: {
    flex: 1,
    paddingVertical: 12,
    borderRadius: 8,
    backgroundColor: colors.gray["02"],
    alignItems: "center",
    justifyContent: "center",
  },
  baseLabel: {
    fontSize: 16,
  },
  btnNormal: {
    backgroundColor: colors.gray["02"],
  },
  txtNormal: {
    color: colors.black,
  },
  btnDanger: {
    backgroundColor: colors.red["02"],
  },
  txtDanger: {
    color: colors.red["07"],
  },
  btnSuccess: {
    backgroundColor: colors.green["02"],
  },
  txtSuccess: {
    color: colors.green["07"],
  },
  btnFocus: {
    backgroundColor: colors.primary,
  },
  txtFocus: {
    color: colors.white,
  },
});
