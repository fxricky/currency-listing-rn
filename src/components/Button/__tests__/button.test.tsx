import { render } from "@testing-library/react-native";
import Button from "../index";

describe("Button", () => {
  test("Normal Button", () => {
    const tree = render(<Button label="Test" onPress={() => {}} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Danger Button", () => {
    const tree = render(
      <Button label="Test" onPress={() => {}} variant="danger" />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Success Button", () => {
    const tree = render(
      <Button label="Test" onPress={() => {}} variant="success" />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Focus Button", () => {
    const tree = render(
      <Button label="Test" onPress={() => {}} variant="focus" />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
