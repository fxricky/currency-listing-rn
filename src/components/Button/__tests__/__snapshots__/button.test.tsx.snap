// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Danger Button 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#f3f4f6",
        "borderRadius": 8,
        "flex": 1,
        "justifyContent": "center",
        "paddingVertical": 12,
      },
      {
        "backgroundColor": "#ffe2e2",
      },
    ]
  }
>
  <Text
    numberOfLines={1}
    style={
      [
        {
          "fontSize": 16,
        },
        {
          "color": "#c10007",
        },
      ]
    }
  >
    Test
  </Text>
</View>
`;

exports[`Button Focus Button 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#f3f4f6",
        "borderRadius": 8,
        "flex": 1,
        "justifyContent": "center",
        "paddingVertical": 12,
      },
      {
        "backgroundColor": "#2b6cb0",
      },
    ]
  }
>
  <Text
    numberOfLines={1}
    style={
      [
        {
          "fontSize": 16,
        },
        {
          "color": "#fff",
        },
      ]
    }
  >
    Test
  </Text>
</View>
`;

exports[`Button Normal Button 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#f3f4f6",
        "borderRadius": 8,
        "flex": 1,
        "justifyContent": "center",
        "paddingVertical": 12,
      },
      {
        "backgroundColor": "#f3f4f6",
      },
    ]
  }
>
  <Text
    numberOfLines={1}
    style={
      [
        {
          "fontSize": 16,
        },
        {
          "color": "#000",
        },
      ]
    }
  >
    Test
  </Text>
</View>
`;

exports[`Button Success Button 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#f3f4f6",
        "borderRadius": 8,
        "flex": 1,
        "justifyContent": "center",
        "paddingVertical": 12,
      },
      {
        "backgroundColor": "#dcfce7",
      },
    ]
  }
>
  <Text
    numberOfLines={1}
    style={
      [
        {
          "fontSize": 16,
        },
        {
          "color": "#008236",
        },
      ]
    }
  >
    Test
  </Text>
</View>
`;
